src/lib/SoftHSM.cpp

```diff
@@ -12830,22 +12831,26 @@ CK_RV SoftHSM::MechParamCheckRSAPKCSOAEP(CK_MECHANISM_PTR pMechanism)
 	return CKR_OK;
 }
 
-bool SoftHSM::isMechanismPermitted(OSObject* key, CK_MECHANISM_PTR pMechanism)
+bool SoftHSM::isMechanismPermitted(OSObject* key, CK_MECHANISM_TYPE mechanism)
 {
 	std::list<CK_MECHANISM_TYPE> mechs = supportedMechanisms;
 	/* First check if the algorithm is enabled in the global configuration */
-	auto it = std::find(mechs.begin(), mechs.end(), pMechanism->mechanism);
+	auto it = std::find(mechs.begin(), mechs.end(), mechanism);
 	if (it == mechs.end())
 		return false;
 
-	OSAttribute attribute = key->getAttribute(CKA_ALLOWED_MECHANISMS);
-	std::set<CK_MECHANISM_TYPE> allowed = attribute.getMechanismTypeSetValue();
+	/* If we have object, consult also its allowed mechanisms */
+	if (key) {
+		OSAttribute attribute = key->getAttribute(CKA_ALLOWED_MECHANISMS);
+		std::set<CK_MECHANISM_TYPE> allowed = attribute.getMechanismTypeSetValue();
 
-	if (allowed.empty()) {
+		if (allowed.empty()) {
```

Comment:
nit: Is this special case needed?

Reply:
Yes. The empty "allowlist" means that everything that is built-in is allowed. The followed check would result in "false" return value for all mechanisms if no allow list would be specified. But this deserves a comment, which I will add here.